File: devicesetup/src/main/java/io/particle/android/sdk/utils/BetterAsyncTaskLoader.java
package io.particle.android.sdk.utils;

import android.content.Context;
import android.support.v4.content.AsyncTaskLoader;


public abstract class BetterAsyncTaskLoader<T> extends AsyncTaskLoader<T> {

    public BetterAsyncTaskLoader(Context context) {
        super(context);
    }

    public abstract boolean hasContent();

    public abstract T getLoadedContent();

    @Override
    protected void onStartLoading() {
        // How is this not on AsyncTaskLoader already?
        if (hasContent()) {
            deliverResult(getLoadedContent());
        }
        if (takeContentChanged() || !hasContent()) {
            forceLoad();
        }
    }
}

File: devicesetup/src/main/java/io/particle/android/sdk/utils/BetterAsyncTaskLoader.kt
package io.particle.android.sdk.utils

import android.content.Context
import android.support.v4.content.AsyncTaskLoader


abstract class BetterAsyncTaskLoader<T>(context: Context) : AsyncTaskLoader<T>(context) {

    abstract val loadedContent: T?

    abstract fun hasContent(): Boolean

    override fun onStartLoading() {
        // How is this not on AsyncTaskLoader already?
        if (hasContent()) {
            deliverResult(loadedContent)
        }
        if (takeContentChanged() || !hasContent()) {
            forceLoad()
        }
    }
}
